File: apps/frontend/src/services/grpc/newsletter.service.ts
/**
 * Newsletter Service gRPC Client
 * Service for handling newsletter subscriptions
 * 
 * @author NyNus Team
 * @version 2.0.0
 */

import { NewsletterServiceClient } from '@/generated/v1/NewsletterServiceClientPb';
import {
  NewsletterSubscribeRequest,
  NewsletterSubscribeResponse,
  NewsletterUnsubscribeRequest,
  NewsletterUnsubscribeResponse,
  ListSubscriptionsRequest,
  ListSubscriptionsResponse,
  GetSubscriptionRequest,
  GetSubscriptionResponse,
  UpdateSubscriptionTagsRequest,
  UpdateSubscriptionTagsResponse,
  DeleteSubscriptionRequest,
  DeleteSubscriptionResponse,
  NewsletterSubscription,
  SubscriptionStats,
} from '@/generated/v1/newsletter_pb';
import { PaginationRequest } from '@/generated/common/common_pb';
import { RpcError } from 'grpc-web';

// gRPC client configuration
const GRPC_ENDPOINT = process.env.NEXT_PUBLIC_GRPC_URL || 'http://localhost:8080';
const newsletterServiceClient = new NewsletterServiceClient(GRPC_ENDPOINT);

// Helper to get auth metadata
function getAuthMetadata(): { [key: string]: string } {
  if (typeof window !== 'undefined') {
    const token = localStorage.getItem('nynus-auth-token');
    if (token) {
      return { 'authorization': `Bearer ${token}` };
    }
  }
  return {};
}

// Handle gRPC errors
function handleGrpcError(error: RpcError): string {
  console.error('gRPC Error:', error);
  switch (error.code) {
    case 3: return error.message || 'Invalid input provided';
    case 6: return 'Email already subscribed';
    case 7: return 'Permission denied';
    case 14: return 'Service temporarily unavailable';
    case 16: return 'Authentication required';
    default: return error.message || 'An unexpected error occurred';
  }
}

// Map NewsletterSubscription from protobuf
function mapSubscriptionFromPb(sub: NewsletterSubscription): any {
  return {
    id: sub.getId(),
    email: sub.getEmail(),
    status: sub.getStatus(),
    subscription_id: sub.getSubscriptionId(),
    confirmed_at: sub.getConfirmedAt()?.toDate(),
    unsubscribed_at: sub.getUnsubscribedAt()?.toDate(),
    unsubscribe_reason: sub.getUnsubscribeReason(),
    source: sub.getSource(),
    tags: sub.getTagsList(),
    metadata: sub.getMetadataMap().toObject(),
    created_at: sub.getCreatedAt()?.toDate(),
    updated_at: sub.getUpdatedAt()?.toDate(),
  };
}

// Map SubscriptionStats from protobuf
function mapStatsFromPb(stats: SubscriptionStats): any {
  return {
    total_active: stats.getTotalActive(),
    total_unsubscribed: stats.getTotalUnsubscribed(),
    total_bounced: stats.getTotalBounced(),
    total_pending: stats.getTotalPending(),
    new_this_week: stats.getNewThisWeek(),
    new_this_month: stats.getNewThisMonth(),
  };
}

export class NewsletterService {
  /**
   * Subscribe to newsletter
   */
  static async subscribe(data: any): Promise<any> {
    try {
      const request = new NewsletterSubscribeRequest();
      request.setEmail(data.email);
      if (data.tags) request.setTagsList(data.tags);
      if (data.metadata) {
        const map = request.getMetadataMap();
        Object.entries(data.metadata).forEach(([key, value]) => {
          map.set(key, value as string);
        });
      }

      // No auth metadata needed for public endpoint
      const response = await newsletterServiceClient.subscribe(request, {});
      const responseObj = response.toObject();
      
      return {
        success: responseObj.response?.success || false,
        message: responseObj.response?.message || '',
        errors: responseObj.response?.errorsList || [],
        subscription: responseObj.subscription ? mapSubscriptionFromPb(response.getSubscription()!) : undefined
      };
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      return {
        success: false,
        message: errorMessage,
        errors: [errorMessage],
        subscription: undefined
      };
    }
  }

  /**
   * Unsubscribe from newsletter
   */
  static async unsubscribe(data: any): Promise<any> {
    try {
      const request = new NewsletterUnsubscribeRequest();
      request.setEmail(data.email);
      if (data.reason) request.setReason(data.reason);

      // No auth metadata needed for public endpoint
      const response = await newsletterServiceClient.unsubscribe(request, {});
      const responseObj = response.toObject();
      
      return {
        success: responseObj.response?.success || false,
        message: responseObj.response?.message || '',
        errors: responseObj.response?.errorsList || []
      };
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      return {
        success: false,
        message: errorMessage,
        errors: [errorMessage]
      };
    }
  }

  /**
   * List subscriptions (Admin)
   */
  static async listSubscriptions(req: any = {}): Promise<any> {
    try {
      const request = new ListSubscriptionsRequest();
      
      // Set pagination
      if (req.pagination) {
        const pagination = new PaginationRequest();
        pagination.setPage(req.pagination.page || 1);
        pagination.setLimit(req.pagination.limit || 20);
        request.setPagination(pagination);
      }
      
      // Set filters
      if (req.status) request.setStatus(req.status);
      if (req.search) request.setSearch(req.search);
      if (req.tags) request.setTagsList(req.tags);

      const response = await newsletterServiceClient.listSubscriptions(request, getAuthMetadata());
      const responseObj = response.toObject();
      
      return {
        success: responseObj.response?.success || false,
        message: responseObj.response?.message || '',
        errors: responseObj.response?.errorsList || [],
        subscriptions: response.getSubscriptionsList().map(mapSubscriptionFromPb),
        pagination: responseObj.pagination,
        stats: responseObj.stats ? mapStatsFromPb(response.getStats()!) : undefined
      };
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      return {
        success: false,
        message: errorMessage,
        errors: [errorMessage],
        subscriptions: [],
        pagination: undefined,
        stats: undefined
      };
    }
  }

  /**
   * Get subscription by email (Admin)
   */
  static async getSubscription(email: string): Promise<any> {
    try {
      const request = new GetSubscriptionRequest();
      request.setEmail(email);

      const response = await newsletterServiceClient.getSubscription(request, getAuthMetadata());
      const responseObj = response.toObject();
      
      return {
        success: responseObj.response?.success || false,
        message: responseObj.response?.message || '',
        errors: responseObj.response?.errorsList || [],
        subscription: responseObj.subscription ? mapSubscriptionFromPb(response.getSubscription()!) : undefined
      };
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      return {
        success: false,
        message: errorMessage,
        errors: [errorMessage],
        subscription: undefined
      };
    }
  }

  /**
   * Update subscription tags (Admin)
   */
  static async updateSubscriptionTags(email: string, tags: string[]): Promise<any> {
    try {
      const request = new UpdateSubscriptionTagsRequest();
      request.setEmail(email);
      request.setTagsList(tags);

      const response = await newsletterServiceClient.updateSubscriptionTags(request, getAuthMetadata());
      const responseObj = response.toObject();
      
      return {
        success: responseObj.response?.success || false,
        message: responseObj.response?.message || '',
        errors: responseObj.response?.errorsList || [],
        subscription: responseObj.subscription ? mapSubscriptionFromPb(response.getSubscription()!) : undefined
      };
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      return {
        success: false,
        message: errorMessage,
        errors: [errorMessage],
        subscription: undefined
      };
    }
  }

  /**
   * Delete subscription (Admin - complete removal)
   */
  static async deleteSubscription(id: string): Promise<any> {
    try {
      const request = new DeleteSubscriptionRequest();
      request.setId(id);

      const response = await newsletterServiceClient.deleteSubscription(request, getAuthMetadata());
      const responseObj = response.toObject();
      
      return {
        success: responseObj.response?.success || false,
        message: responseObj.response?.message || '',
        errors: responseObj.response?.errorsList || []
      };
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      return {
        success: false,
        message: errorMessage,
        errors: [errorMessage]
      };
    }
  }
}

File: apps/backend/internal/grpc/newsletter_service.go
package grpc

import (
	"context"
	"fmt"

	"exam-bank-system/apps/backend/internal/entity"
	"exam-bank-system/apps/backend/internal/service/content/newsletter"
	"exam-bank-system/apps/backend/pkg/proto/common"
	v1 "exam-bank-system/apps/backend/pkg/proto/v1"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// NewsletterServiceServer implements the NewsletterService gRPC service
type NewsletterServiceServer struct {
	v1.UnimplementedNewsletterServiceServer
	newsletterMgmt *newsletter.NewsletterMgmt
}

// NewNewsletterServiceServer creates a new NewsletterService server
func NewNewsletterServiceServer(newsletterMgmt *newsletter.NewsletterMgmt) *NewsletterServiceServer {
	return &NewsletterServiceServer{
		newsletterMgmt: newsletterMgmt,
	}
}

// Subscribe handles newsletter subscription (public endpoint)
func (s *NewsletterServiceServer) Subscribe(ctx context.Context, req *v1.NewsletterSubscribeRequest) (*v1.NewsletterSubscribeResponse, error) {
	// Extract metadata for IP address
	md, _ := metadata.FromIncomingContext(ctx)

	// Create request for management service
	mgmtRequest := &newsletter.SubscribeRequest{
		Email:    req.Email,
		Tags:     req.Tags,
		Metadata: make(map[string]interface{}),
	}

	// Convert metadata map
	if req.Metadata != nil {
		for k, v := range req.Metadata {
			mgmtRequest.Metadata[k] = v
		}
	}

	// Extract IP address from metadata
	if forwardedFor := md.Get("x-forwarded-for"); len(forwardedFor) > 0 {
		mgmtRequest.IPAddress = forwardedFor[0]
	} else if realIP := md.Get("x-real-ip"); len(realIP) > 0 {
		mgmtRequest.IPAddress = realIP[0]
	}

	// Subscribe via management service
	subscription, err := s.newsletterMgmt.Subscribe(mgmtRequest)
	if err != nil {
		return nil, status.Errorf(codes.InvalidArgument, "failed to subscribe: %v", err)
	}

	// Convert to proto response
	response := &v1.NewsletterSubscribeResponse{
		Response: &common.Response{
			Success: true,
			Message: "Successfully subscribed to newsletter! Please check your email for confirmation.",
		},
		Subscription: s.entityToProto(subscription),
	}

	return response, nil
}

// Unsubscribe handles newsletter unsubscription (public endpoint)
func (s *NewsletterServiceServer) Unsubscribe(ctx context.Context, req *v1.NewsletterUnsubscribeRequest) (*v1.NewsletterUnsubscribeResponse, error) {
	// Unsubscribe via management service
	if err := s.newsletterMgmt.Unsubscribe(req.Email, req.Reason); err != nil {
		return nil, status.Errorf(codes.InvalidArgument, "failed to unsubscribe: %v", err)
	}

	response := &v1.NewsletterUnsubscribeResponse{
		Response: &common.Response{
			Success: true,
			Message: "Successfully unsubscribed from newsletter. We're sorry to see you go!",
		},
	}

	return response, nil
}

// ListSubscriptions retrieves newsletter subscriptions (admin only)
func (s *NewsletterServiceServer) ListSubscriptions(ctx context.Context, req *v1.ListSubscriptionsRequest) (*v1.ListSubscriptionsResponse, error) {
	// Parse pagination
	page := int32(1)
	pageSize := int32(20)
	if req.Pagination != nil {
		if req.Pagination.Page > 0 {
			page = req.Pagination.Page
		}
		if req.Pagination.Limit > 0 && req.Pagination.Limit <= 100 {
			pageSize = req.Pagination.Limit
		}
	}

	// Get subscriptions from management service
	subscriptions, totalCount, err := s.newsletterMgmt.ListSubscriptions(
		req.Status,
		req.Search,
		req.Tags,
		int(page),
		int(pageSize),
	)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to list subscriptions: %v", err)
	}

	// Get statistics
	stats, err := s.newsletterMgmt.GetStats()
	if err != nil {
		// Don't fail the request if stats fail
		stats = &newsletter.SubscriptionStats{}
	}

	// Convert to proto
	protoSubscriptions := make([]*v1.NewsletterSubscription, len(subscriptions))
	for i, subscription := range subscriptions {
		protoSubscriptions[i] = s.entityToProto(subscription)
	}

	// Calculate total pages
	totalPages := (totalCount + int(pageSize) - 1) / int(pageSize)

	response := &v1.ListSubscriptionsResponse{
		Response: &common.Response{
			Success: true,
			Message: fmt.Sprintf("Found %d subscriptions", totalCount),
		},
		Subscriptions: protoSubscriptions,
		Pagination: &common.PaginationResponse{
			Page:       page,
			Limit:      pageSize,
			TotalCount: int32(totalCount),
			TotalPages: int32(totalPages),
		},
		Stats: &v1.SubscriptionStats{
			TotalActive:       int32(stats.TotalActive),
			TotalUnsubscribed: int32(stats.TotalUnsubscribed),
			TotalBounced:      int32(stats.TotalBounced),
			TotalPending:      int32(stats.TotalPending),
			NewThisWeek:       int32(stats.NewThisWeek),
			NewThisMonth:      int32(stats.NewThisMonth),
		},
	}

	return response, nil
}

// GetSubscription retrieves a specific subscription (admin only)
func (s *NewsletterServiceServer) GetSubscription(ctx context.Context, req *v1.GetSubscriptionRequest) (*v1.GetSubscriptionResponse, error) {
	// Get subscription from management service
	subscription, err := s.newsletterMgmt.GetSubscription(req.Email)
	if err != nil {
		return nil, status.Errorf(codes.NotFound, "subscription not found: %v", err)
	}

	response := &v1.GetSubscriptionResponse{
		Response: &common.Response{
			Success: true,
			Message: "Subscription retrieved successfully",
		},
		Subscription: s.entityToProto(subscription),
	}

	return response, nil
}

// UpdateSubscriptionTags updates tags for a subscription (admin only)
func (s *NewsletterServiceServer) UpdateSubscriptionTags(ctx context.Context, req *v1.UpdateSubscriptionTagsRequest) (*v1.UpdateSubscriptionTagsResponse, error) {
	// Update tags via management service
	subscription, err := s.newsletterMgmt.UpdateTags(req.Email, req.Tags)
	if err != nil {
		return nil, status.Errorf(codes.InvalidArgument, "failed to update tags: %v", err)
	}

	response := &v1.UpdateSubscriptionTagsResponse{
		Response: &common.Response{
			Success: true,
			Message: "Tags updated successfully",
		},
		Subscription: s.entityToProto(subscription),
	}

	return response, nil
}

// DeleteSubscription removes a subscription completely (admin only)
func (s *NewsletterServiceServer) DeleteSubscription(ctx context.Context, req *v1.DeleteSubscriptionRequest) (*v1.DeleteSubscriptionResponse, error) {
	// Delete subscription via management service
	if err := s.newsletterMgmt.DeleteSubscription(req.Id); err != nil {
		return nil, status.Errorf(codes.NotFound, "failed to delete subscription: %v", err)
	}

	response := &v1.DeleteSubscriptionResponse{
		Response: &common.Response{
			Success: true,
			Message: "Subscription deleted successfully",
		},
	}

	return response, nil
}

// Helper function to convert entity to proto
func (s *NewsletterServiceServer) entityToProto(e *entity.NewsletterSubscription) *v1.NewsletterSubscription {
	proto := &v1.NewsletterSubscription{
		Id:             e.ID.String(),
		Email:          e.Email,
		Status:         string(e.Status),
		SubscriptionId: e.SubscriptionID,
		Source:         e.Source,
		Tags:           e.Tags,
		Metadata:       make(map[string]string),
		CreatedAt:      timestamppb.New(e.CreatedAt),
		UpdatedAt:      timestamppb.New(e.UpdatedAt),
	}

	// Convert optional fields
	if e.ConfirmedAt != nil {
		proto.ConfirmedAt = timestamppb.New(*e.ConfirmedAt)
	}
	if e.UnsubscribedAt != nil {
		proto.UnsubscribedAt = timestamppb.New(*e.UnsubscribedAt)
	}
	if e.UnsubscribeReason != nil {
		proto.UnsubscribeReason = *e.UnsubscribeReason
	}

	// Convert metadata map
	if e.Metadata != nil {
		for k, v := range e.Metadata {
			proto.Metadata[k] = fmt.Sprintf("%v", v)
		}
	}

	return proto
}

